src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets

```diff
@@ -2058,7 +2060,28 @@ because xbuild doesn't support framework reference assemblies.
 </Target>
 	
 <Target Name="_LinkAssemblies"
-  DependsOnTargets="_ResolveAssemblies;_CreatePackageWorkspace;_LinkAssembliesNoShrink;_LinkAssembliesShrink" />
+  DependsOnTargets="_ResolveAssemblies;_CreatePackageWorkspace;_GenerateJniMarshalMethods;_LinkAssembliesNoShrink;_LinkAssembliesShrink" />
+
+<Target Name="_GenerateJniMarshalMethods"
+  Condition="'$(AndroidGenerateJniMarshalMethods)' == 'True' And '$(AndroidLinkMode)' != 'None'"
+  DependsOnTargets="_GetReferenceAssemblyPaths"
+  Inputs="@(ResolvedUserAssemblies->'$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)')"
+  Outputs="$(_AndroidJniMarshalMethodsFlag)">
+
+    <JdkInfo>
+      <Output TaskParameter="JdkJvmPath" PropertyName="_AndroidJvmPath" />
+    </JdkInfo>
+
+    <Exec
+       Command="MONO_PATH=$(_XATargetFrameworkDirectories) $(MonoAndroidBinDirectory)\mono $(MonoAndroidBinDirectory)\..\jnimarshalmethod-gen.exe --jvm=$(_AndroidJvmPath) @(ResolvedUserAssemblies->'$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)', ' ')"
```

Comment:
I suspect that this should be quoted: `MONO_PATH=&quot;$(_XATargetFrameworkDirectories)&quot; ...`.

This should use `$(MonoAndroidToolsDirectory)` instead of `$(MonoAndroidBinDirectory)\..`.

All parameters must be properly quoted, *especially* the `@(ResolvedUserAssemblies->...)` output, as those files are within the project directory, which could contain spaces.